internal/framework/fetch/fetch.go

```diff
@@ -0,0 +1,569 @@
+}
+
+// S3Error represents an error when fetching from S3 fails.
+type S3Error struct {
+	Err    error
+	Bucket string
+	Key    string
+}
+
+func (e *S3Error) Error() string {
+	return fmt.Sprintf("S3 error for s3://%s/%s: %v", e.Bucket, e.Key, e.Err)
+}
+
+func (e *S3Error) Unwrap() error {
+	return e.Err
+}
+
+// HTTPStatusError represents an error for an unexpected HTTP status code.
+type HTTPStatusError struct {
+	StatusCode int
+}
+
+func (e *HTTPStatusError) Error() string {
+	return fmt.Sprintf("unexpected status code: %d", e.StatusCode)
+}
+
+// HTTPError represents an error when fetching via HTTP fails.
+type HTTPError struct {
+	Err error
+	URL string
+}
+
+func (e *HTTPError) Error() string {
+	return fmt.Sprintf("HTTP error for %s: %v", e.URL, e.Err)
+}
+
+func (e *HTTPError) Unwrap() error {
+	return e.Err
+}
+
+// RetryBackoffType defines supported backoff strategies.
+type RetryBackoffType string
+
+const (
+	RetryBackoffExponential RetryBackoffType = "exponential"
+	RetryBackoffLinear      RetryBackoffType = "linear"
+)
+
+// options contains the configuration for fetching remote files.
+type options struct {
+	checksumLocation string
+	retryBackoff     RetryBackoffType
+	timeout          time.Duration
+	retryMaxDelay    time.Duration
+	retryAttempts    int32
+	checksumEnabled  bool
+}
+
+// defaults returns options with sensible default values.
+func defaults() options {
+	return options{
+		timeout:       defaultTimeout,
+		retryAttempts: defaultRetryAttempts,
+		retryMaxDelay: defaultRetryMaxDelay,
+		retryBackoff:  RetryBackoffExponential,
+	}
+}
+
+// Option defines a function that modifies fetch options.
+type Option func(*options)
+
+// WithTimeout sets the HTTP request timeout.
+func WithTimeout(timeout time.Duration) Option {
+	return func(o *options) {
+		o.timeout = timeout
+	}
+}
+
+// WithRetryAttempts sets the number of retry attempts (total attempts = 1 + retries).
+func WithRetryAttempts(attempts int32) Option {
+	return func(o *options) {
+		o.retryAttempts = attempts
+	}
+}
+
+// WithRetryBackoff sets the retry backoff strategy.
+func WithRetryBackoff(backoff RetryBackoffType) Option {
+	return func(o *options) {
+		o.retryBackoff = backoff
+	}
+}
+
+// WithMaxRetryDelay sets the maximum delay between retries.
+func WithMaxRetryDelay(delay time.Duration) Option {
+	return func(o *options) {
+		o.retryMaxDelay = delay
+	}
+}
+
+// WithChecksum enables checksum validation with an optional custom checksum location.
+// For HTTP URLs: if no location is provided, defaults to <fileURL>.sha256
+// For S3 URLs: if no location is provided, defaults to <key>.sha256 in the same bucket.
+func WithChecksum(checksumLocation ...string) Option {
+	return func(o *options) {
+		o.checksumEnabled = true
+		if len(checksumLocation) > 0 {
+			o.checksumLocation = checksumLocation[0]
+		}
+	}
+}
+
+// S3Client defines the interface for S3 operations.
+//
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . S3Client
```

Comment:
Nowadays you can just specify `//go:generate go tool counterfeiter -generate` once per package, and then wherever you need it, set `//counterfeiter:generate . <interface-name>`.